lib/spack/spack/spec.py

```diff
@@ -343,17 +371,25 @@ def satisfies(self, other, strict=False):
                        for attr in odict if sdict[attr] and odict[attr])
 
     def constrain(self, other):
-        """ Projects all architecture fields that are specified in the given
-            spec onto the instance spec if they're missing from the instance
-            spec. This will only work if the two specs are compatible.
+        """Projects all architecture fields that are specified in the given
+        spec onto the instance spec if they're missing from the instance
+        spec.
+
+        This will only work if the two specs are compatible.
```

Comment:
I think what should ultimately happen here is that if you constrain a `haswell` spec by `icelake` that works, but the other way around does not (it's already too specific).

Does it make sense to implement that now?  While we may not currently support *matching* existing installs with compatible architectures when reusing specs, i think implementing `constrain()` and `satisfies()` correctly can still work well in concretization.  As in, Spack should allow you to build against more generic binaries if you are explicit about it, kind of like we currently do with compilers.

Reply:
We could probably defer this to a later PR, as it deals with the niche case of mixing specs for different targets in the same DAG. Or am I getting the comment wrong?

Reply:
I think deferring this is fine.  I don't think it's as niche as you might think -- it's likely when we start distributing binaries that certain binaries (ones that are not performance sensitive) will be built for a more generic architecture than the ones that really benefit from optimization.